File: calc.clj
#!/usr/bin/env -S clj -M

(loop []
	(loop [s [] toks (re-seq #"\S+" (read-line))]
		(if-let [t (first toks)]
			(if-let [n (parse-long t)]
				(recur (conj s n) (rest toks))
				(if-let [[b a] (take-last 2 s)]
					(if-let [op ({"+" + "-" - "*" * "/" quot} t)]
						(recur (conj (drop-last 2 s) (long (op a b))) (rest toks))
						(println "Invalid Token"))
					(println "Too Few Arguments")))
			(println (if (= 1 (count s)) (first s) "Invalid Input"))))
	(recur))

File: calc.ml
#!/usr/bin/env ocaml

let parse = function
  | "+" -> Some (`Op ( + ))
  | "-" -> Some (`Op ( - ))
  | "*" -> Some (`Op ( * ))
  | "/" -> Some (`Op ( / ))
  | s -> s |> int_of_string_opt |> Option.map (fun n -> `Num n)

let eval item nums =
  let apply op = function
    | b :: a :: nums -> Ok (op a b :: nums)
    | _ -> Error "Too Few Arguments"
  in
  match parse item with
  | Some (`Op op) -> apply op nums
  | Some (`Num n) -> Ok (n :: nums)
  | None -> Error "Invalid Token"

let eval nums item = Result.bind nums (eval item)

let run line =
  match line |> String.split_on_char ' ' |> List.fold_left eval (Ok []) with
  | Ok [ x ] -> Int.to_string x
  | Error e -> e
  | _ -> "Invalid Input"

let read () =
  match read_line () with
  | s -> Some (String.trim s)
  | exception Sys.Break -> None

let () =
  Sys.catch_break true;
  Seq.of_dispenser read |> Seq.map run |> Seq.iter print_endline
